src/outputs/parthenon_hdf5.cpp

```diff
@@ -192,436 +330,652 @@ void PHDF5Output::genXDMF(std::string hdfFile, Mesh *pm, SimTime *tm) {
+  // of the sparse fields (field name with sparse id, i.e. "bar_28", "bar_7", foo_1",
+  // "foo_145"). The field names are in alphabetical order, which is the same order
+  // they show up in all_unique_vars (because it's a sorted set).
+  //
+  // The dataset SparseInfo itself is a 2D array of bools. The first index is the
+  // global block index and the second index is the sparse field (same order as the
+  // SparseFields attribute). SparseInfo[b][v] is true if the sparse field with index
+  // v is expanded on the block with index b, otherwise the value is false
+
+  std::vector<std::string> sparse_names;
+  std::unordered_map<std::string, size_t> sparse_field_idx;
+  for (auto &vinfo : all_unique_vars) {
+    if (vinfo.is_sparse) {
+      sparse_field_idx.insert({vinfo.label, sparse_names.size()});
+      sparse_names.push_back(vinfo.label);
+    }
+  }
+
+  hsize_t num_sparse = sparse_names.size();
+  // can't use std::vector here because std::vector<hbool_t> is the same as
+  // std::vector<bool> and it doesn't have .data() member
+  std::unique_ptr<hbool_t[]> sparse_expanded(new hbool_t[num_blocks_local * num_sparse]);
+
+  // allocate space for largest size variable
+  const hsize_t varSize = nx3 * nx2 * nx1;
+  int vlen_max = 0;
+  for (auto &vinfo : all_unique_vars) {
+    vlen_max = std::max(vlen_max, vinfo.vlen);
+  }
+
+  using OutT = typename std::conditional<WRITE_SINGLE_PRECISION, float, Real>::type;
+  std::vector<OutT> tmpData(varSize * vlen_max * num_blocks_local);
+
+  // create persistent spaces
+  local_count[0] = num_blocks_local;
+  global_count[0] = max_blocks_global;
+  local_count[1] = global_count[1] = nx3;
+  local_count[2] = global_count[2] = nx2;
+  local_count[3] = global_count[3] = nx1;
+
+  // for each variable we write
+  for (auto &vinfo : all_unique_vars) {
+    // not really necessary, but doesn't hurt
+    memset(tmpData.data(), 0, tmpData.size() * sizeof(OutT));
+
+    const std::string var_name = vinfo.label;
+    const hsize_t vlen = vinfo.vlen;
+
+    local_count[4] = global_count[4] = vlen;
+
+    // load up data
+    hsize_t index = 0;
+    bool found_any = false;
+
+    // for each local mesh block
+    for (size_t b_idx = 0; b_idx < num_blocks_local; ++b_idx) {
+      const auto &pmb = pm->block_list[b_idx];
+      bool found = false;
+
+      // for each variable that this local meshblock actually has
+      auto ci = get_MeshBlockDataIterator(pmb);
+      for (auto &v : ci.vars) {
+        // Note index l transposed to interior
+        if (var_name.compare(v->label()) == 0) {
+          auto v_h = v->data.GetHostMirrorAndCopy();
+          for (int k = out_kb.s; k <= out_kb.e; ++k) {
+            for (int j = out_jb.s; j <= out_jb.e; ++j) {
+              for (int i = out_ib.s; i <= out_ib.e; ++i) {
+                for (int l = 0; l < vlen; ++l) {
+                  tmpData[index++] = static_cast<OutT>(v_h(l, k, j, i));
+                }
+              }
+            }
           }
```

Comment:
I don't love this loop being done on the host. I wonder if there's a way to do it on device. We could do the index math ahead of time, then in a kernel copy to a temporary view, that's then copied to host in a single htod instad of nvars htods?

Reply:
Hmm, interesting suggestion. In order to keep memory requirements manageable, I guess we would need to launch one kernel per meshblock, right? That would come with the complication of moving the mesh-block specific list of variables to the device. But I guess that would just be a variable pack per meshblock, although it would potentially have to be rebuilt each time because which sparse ids are allocated on a particular block can change at any time.

It's not obvious to me that this would actually improve performance significantly. But I guess we'd have to try to find out. I'm not sure if I'd want to hold up this PR (and thus the sparse implementation) to find out what the performance difference would be, though.

Reply:
I've made issue #516 for this, since this PR needs to move along and I need to focus on actual sparse implementation first.

Reply:
That seems reasonable